docs: Provide Design document for HIP-904 System Contracts flows (#15435)
hashgraph/hedera-services: hedera-node/docs/design/services/smart-contract-service/frictionless-airdrops-system-contracts.md

## Goals

- Expose `airdropToken`, `claimAirdrops`, `cancelAirdrops` and `rejectTokens` as new functions in the Hedera Token Service Smart Contract.
- Expose `cancelAirdropFT`, `cancleAirdropNFT`, `claimAirdropFT`, `claimAirdropNFT`, `rejectTokenFt`, `rejectTokenNFTs`, `setAutomaticAssociations` as new functions in the proxy redirect token facade contract IHRC.
- Implement the needed HTS system contract classes to support the new functions.

## Non Goals

- The implementation of the HAPI operation, as it is already an existing feature.

## Architecture

The architecture for the frictionless airdrops follows the existing framework defined for handling all calls to the HederaTokenService system contract in the modularization services and is described in more detail in the Implementation section below.

## Implementation

### New Solidity Structures

We will introduce the following new structures to support the new functionality:

`PendingAirdrop` - A struct that represents a pending airdrop request.

```solidity
struct PendingAirdrop {
    address sender;
    address receiver;

    address token;

>     bool isNft;

Reviewer: Why not an enum for (fungible, non-fungible) instead of this `bool isNft`, then put it _above_ the `address token`.  Reason would be to look more similar to the protobuf `PendingAirdrop`.

    int64 serial;
}
```

> `NftId` - A struct that represents the Nft serials to be rejected.

Reviewer: Also is `NftId` the right name - there are multiple Nfts you can identify by their serial.  So, plural vs singular?  Doesn't really match the HIP either which has a different defintiion of `NftId` in the protobuf - a singular tokentype+serial number.

```solidity
struct NftId {
    address nft;
    int64[] serials;
}
```

### New Solidity Functions

New system contract functions must be added to the `IHederaTokenService` interface to support airdropping tokens.

| Function Selector Hash |                                                 Function Signature                                                  |    Response    |                                 |
|------------------------|---------------------------------------------------------------------------------------------------------------------|----------------|---------------------------------|
| `0x2f348119`           | `function airdropTokens(TokenTransferList[] memory tokenTransfers) external returns (int64 responseCode)`           | `ResponseCode` | The response code from the call |
| `0xc1acfe5d`           | `function cancelAirdrops(PendingAirdrop[] memory pendingAirdrops) external returns (int64 responseCode)`            | `ResponseCode` | The response code from the call |

> | `0x241b6af7`           | `function claimAirdrops(PendingAirdrop[] memory pendingAirdrops) external returns (int64 responseCode)`             | `ResponseCode` | The response code from the call |

Reviewer: Here (and in next table) the HIP limits the number of pending airdrops to a maximum of 10.  That limit should be in this document somewhere - perhaps in this table, perhaps as text following the table?

> | `0x012ea0b1`           | `function rejectTokens(address[] memory ftAddresses,  NftId[] memory nftIds) external returns (int64 responseCode)` | `ResponseCode` | The response code from the call |

Reviewer: (Oddly, `rejectTokens` does _not_ have a limit (in the HIP) on how many you can reject at one time.  Not sure why you're limited to 10 for cancel but no limit for reject - you'd think similar limits on resource consumption in a single transaction would apply.)

| `0xcef5b705`           | `function cancelAirdropFT(address receiverAddress) external returns (uint256 responseCode)`                      | `ResponseCode` | The response code from the call |
| `0xad4917cf`           | `function cancelAirdropNFT(address receiverAddress, int64 serialNumber) external returns (uint256 responseCode)` | `ResponseCode` | The response code from the call |
| `0xa83bc5b2`           | `function claimAirdropFT(address senderAddress) external returns (uint256 responseCode)`                         | `ResponseCode` | The response code from the call |
| `0x63ada5d7`           | `function claimAirdropNFT(address senderAddress, int64 serialNumber) external returns (uint256 responseCode)`    | `ResponseCode` | The response code from the call |
| `0x76c6b391`           | `function rejectTokenFT() external returns (uint256 responseCode)`                                               | `ResponseCode` | The response code from the call |
| `0xa869c78a`           | `function rejectTokenNFTs(int64[] memory serialNumbers) external returns (uint256 responseCode)`                 | `ResponseCode` | The response code from the call |

> | `0x966884d4`           | `function setAutomaticAssociations(int64 newAutoAssociations) external returns (uint256 responseCode)`           | `ResponseCode` | The response code from the call |

Reviewer: This is HAS not HTS right?  Does that need to be called out?

# Frictionless Airdrops via System Contracts

## Purpose

[HIP-904](https://hips.hedera.com/hip/hip-904) introduced the Frictionless Airdrops feature for fungible and non-fungible tokens.
This document will define the architecture and implementation of the `airdropToken`, `claimAirdrops`, `cancelAirdrops` and `rejectTokens` smart contract functions
and their respective redirect function calls (`cancelAirdropFT`, `cancleAirdropNFT`, `claimAirdropFT`, `claimAirdropNFT`, `rejectTokenFT`, `rejectTokensNFT`, `setAutomaticAssociations`) that will extend the capabilities of the Hedera Smart Contract Service (HSCS) to support frictionless airdrops.

## References

[HIP-904](https://hips.hedera.com/hip/hip-904) - HIP that introduces the frictionless airdrops.

## Goals

- Expose `airdropToken`, `claimAirdrops`, `cancelAirdrops` and `rejectTokens` as new functions in the Hedera Token Service Smart Contract.
- Expose `cancelAirdropFT`, `cancleAirdropNFT`, `claimAirdropFT`, `claimAirdropNFT`, `rejectTokenFt`, `rejectTokensNFT`, `setAutomaticAssociations` as new functions in the proxy redirect token facade contract IHRC.

> - Implement the needed HTS system contract classes to support the new functions.

Reviewer 2: `getPendingAirdrops` is not mentioned in the goals.

Author: Thanks, will add.

## Non Goals

- The implementation of the HAPI operation, as it is already an existing feature.

## Architecture

The architecture for the frictionless airdrops follows the existing framework defined for handling all calls to the HederaTokenService system contract in the modularization services and is described in more detail in the Implementation section below.

## Implementation

### New Solidity Structures

We will introduce the following new structures to support the new functionality:

`PendingAirdrop` - A struct that represents a pending airdrop request.

```solidity
struct PendingAirdrop {
    address sender;
    address receiver;

    address token;
    bool isNft;
    int64 serial;
}
```

`PendingAirdropRecord` - A struct that represents a pending airdrop record.

```solidity
struct PendingAirdropRecord {
    PendingAirdrop pendingAirdrop;
    uint64 amount;
}
```

`NftId` - A struct that represents the Nft serials to be rejected.

```solidity

> struct NftId {

Reviewer 2: Can this struct be used in other places than just related to the airdrop functionality?

Author: I think we can incorporate it with functions that manipulate/read data for multiple NFTs with multiple serials.

- Verify that the `cancelAirdropFT` function cancels a pending airdrop of the redirected token.
- Verify that the `cancelAirdropNFT` function cancels a pending airdrop of the redirected nft serial.
- Verify that the `claimAirdropFT` function claims a pending airdrop of the redirected token.
- Verify that the `claimAirdropNFT` function claims a pending airdrop of the redirected nft serial number.
- Verify that the `rejectTokenFT` function rjects tokens for a given account.
- Verify that the `rejectTokensNFT` function rejects tokens for a given account and serial number.
- Verify that the `setAutomaticAssociations` function sets the automatic associations for a given account.
- Verify that the `setAutomaticAssociations` function sets the automatic associations to unlimited (-1) for a given account.
- Verify that the `setAutomaticAssociations` function sets the automatic associations to zero for a given account.

#### Negative Tests

- Verify that the `airdropTokens` function fails when the sender does not have enough balance.
- Verify that the `airdropTokens` function fails when the receiver does not have a valid account.

> - Verify that the `airdropTokens` function fails when the token does not exist.

Reviewer 2: Also validate that out of bounds values for the transfer list fails

Author: Thanks, I'll add more negative cases for out of bounds values

| `0x2f348119`           | `function airdropTokens(TokenTransferList[] memory tokenTransfers) external returns (int64 responseCode)`                           | `ResponseCode`                           | The response code from the call                                           |
| `0xc1acfe5d`           | `function cancelAirdrops(PendingAirdrop[] memory pendingAirdrops) external returns (int64 responseCode)`                            | `ResponseCode`                           | The response code from the call                                           |
| `0x241b6af7`           | `function claimAirdrops(PendingAirdrop[] memory pendingAirdrops) external returns (int64 responseCode)`                             | `ResponseCode`                           | The response code from the call                                           |
| `0x012ea0b1`           | `function rejectTokens(address[] memory ftAddresses,  NftId[] memory nftIds) external returns (int64 responseCode)`                 | `ResponseCode`                           | The response code from the call                                           |

> | `0xc60ffe1b`           | `function getPendingAirdrops(address account) external returns (int64 responseCode, PendingAirdropRecord[] memory pendingAirdrops)` | `ResponseCode`, `PendingAirdropRecord[]` | The response code from the call and the pending airdrops for given sender |

Reviewer 3: I don't think we should be adding this function in its form.
How would we handle when the number of airdrops is significant? We can' sent it all and if you send a partial amount how does a user get this information

Author: I see your point. This one is a proposal, as we only have the Mirror node to query for the Pending Airdrops where we have pagination. Maybe we should skip the query part and leave it to the Mirror node for now?

Reviewer: I wonder what the use case is for a smart contract to call this (as opposed to a wallet or something else external to the blockchain) calling the mirror node)?

|------------------------|------------------------------------------------------------------------------------------------------------------|----------------|---------------------------------|
| `0xcef5b705`           | `function cancelAirdropFT(address receiverAddress) external returns (uint256 responseCode)`                      | `ResponseCode` | The response code from the call |
| `0xad4917cf`           | `function cancelAirdropNFT(address receiverAddress, int64 serialNumber) external returns (uint256 responseCode)` | `ResponseCode` | The response code from the call |
| `0xa83bc5b2`           | `function claimAirdropFT(address senderAddress) external returns (uint256 responseCode)`                         | `ResponseCode` | The response code from the call |
| `0x63ada5d7`           | `function claimAirdropNFT(address senderAddress, int64 serialNumber) external returns (uint256 responseCode)`    | `ResponseCode` | The response code from the call |
| `0x76c6b391`           | `function rejectTokenFT() external returns (uint256 responseCode)`                                               | `ResponseCode` | The response code from the call |

> | `0xedd9c546`           | `function rejectTokensNFT(int64[] memory serialNumber) external returns (uint256 responseCode)`                  | `ResponseCode` | The response code from the call |

Reviewer 3: Q: Should this be as follows?
```suggestion
| `0xedd9c546`           | `function rejectTokenNFTs(int64[] memory serialNumbers) external returns (uint256 responseCode)`                  | `ResponseCode` | The response code from the call |
```

Author: Yes, Thanks! This was the way it was described in the HIP, but it definitely reads better this way.

Reviewer 3: Okay, we should correct it, I seem to have missed it so the above is a correction on my own work :)

- Expose `cancelAirdropFT`, `cancleAirdropNFT`, `claimAirdropFT`, `claimAirdropNFT`, `rejectTokenFt`, `rejectTokenNFTs`, `setAutomaticAssociations` as new functions in a new proxy redirect token facade contract IHRC904.
- Implement the needed HTS system contract classes to support the new functions.

## Non Goals

- The implementation of the HAPI operation, as it is already an existing feature.

## Architecture

The architecture for the frictionless airdrops follows the existing framework defined for handling all calls to the HederaTokenService system contract in the modularization services and is described in more detail in the Implementation section below.

## Implementation

### New Solidity Structures

We will introduce the following new structures to support the new functionality:

`PendingAirdrop` - A struct that represents a pending airdrop request.

```solidity
struct PendingAirdrop {
    address sender;
    address receiver;

    address token;
    int64 serial;
}
```

`NftId` - A struct that represents the Nft serials to be rejected.

```solidity
struct NftId {
    address nft;

>     int64[] serials;

Reviewer 3: In follow up from note above this should be as follows in alignment with protobuf.
```suggestion
    int64 serial;
```

| `0xad4917cf`           | `function cancelAirdropNFT(address receiverAddress, int64 serialNumber) external returns (uint256 responseCode)` | TokenCancel      | HTS                 | `ResponseCode` | The response code from the call |
| `0xa83bc5b2`           | `function claimAirdropFT(address senderAddress) external returns (uint256 responseCode)`                         | TokenClaim       | HTS                 | `ResponseCode` | The response code from the call |
| `0x63ada5d7`           | `function claimAirdropNFT(address senderAddress, int64 serialNumber) external returns (uint256 responseCode)`    | TokenClaim       | HTS                 | `ResponseCode` | The response code from the call |
| `0x76c6b391`           | `function rejectTokenFT() external returns (uint256 responseCode)`                                               | TokenReject      | HTS                 | `ResponseCode` | The response code from the call |
| `0xa869c78a`           | `function rejectTokenNFTs(int64[] memory serialNumbers) external returns (uint256 responseCode)`                 | TokenReject      | HTS                 | `ResponseCode` | The response code from the call |

> | `0x966884d4`           | `function setAutomaticAssociations(boolean enableAutoAssociations) external returns (uint256 responseCode)`      | CryptoUpdate     | HAS                 | `ResponseCode` | The response code from the call |

Reviewer 3: Q: should we be more explicit so that the function name signifies the binary nature of the set?
Maybe
```suggestion
| `0x966884d4`           | `function setUnlimitedAutomaticAssociations(boolean enableAutoAssociations) external returns (uint256 responseCode)`      | CryptoUpdate     | HAS                 | `ResponseCode` | The response code from the call |
```

Author: Agreed. This will improve the readability a lot.

- `CancelAirdropsDecoder` - This class provides methods and constants for decoding the given `HtsCallAttempt` into a `PendingAirdropId` list for `TokenCancelAirdrop` call.
- `ClaimAirdropsTranslator` - This class will be responsible for handling the `claimAirdrops`, `claimAirdropFT` and `claimAirdropNFT` selectors and dispatching them to the corresponding HAPI calls.
- `ClaimAirdropsDecoder` - This class provides methods and constants for decoding the given `HtsCallAttempt` into a `PendingAirdropId` list for `TokenClaimAirdrop` call.
- `RejectTokensTranslator` - This class will be responsible for handling the `rejectTokens`, `rejectTokenFT` and `rejectTokenNFT` selectors and dispatching them to the corresponding HAPI calls.
- `RejectTokensDecoder` - This class provides methods and constants for decoding the given `HtsCallAttempt` into `TokenReference` list for `TokenReject` call.
- `SetAutomaticAssociationsTranslator` - This class will be responsible for handling the `setAutomaticAssociations` selector and dispatching it to the corresponding HAPI calls.
- `SetAutomaticAssociationsCall` - This class provides methods for preparing and executing the given `HasCallAttempt`.

> 

Reviewer 3: Please add a config section highlights 5 feature gates (`.tokenAirdrop`, `.tokenCancelAirdrop`, `.tokenClaimAirdrop`, `.tokenReject` and `.setUnlimitedAutomaticAssociations`) that may be used to enable these features and help stage the releases across milestone as needed

Author: Added the section. Thanks

# Frictionless Airdrops via System Contracts

## Purpose

[HIP-904](https://hips.hedera.com/hip/hip-904) introduced the Frictionless Airdrops feature for fungible and non-fungible tokens.
This document will define the architecture and implementation of the `airdropToken`, `claimAirdrops`, `cancelAirdrops` and `rejectTokens` smart contract functions

> and their respective redirect function calls (`cancelAirdropFT`, `cancleAirdropNFT`, `claimAirdropFT`, `claimAirdropNFT`, `rejectTokenFT`, `rejectTokenNFTs`, `setUnlimitedAutomaticAssociations`) that will extend the capabilities of the Hedera Smart Contract Service (HSCS) to support frictionless airdrops.

Reviewer 4: ```suggestion
and their respective redirect function calls (`cancelAirdropFT`, `cancelAirdropNFT`, `claimAirdropFT`, `claimAirdropNFT`, `rejectTokenFT`, `rejectTokenNFTs`, `setUnlimitedAutomaticAssociations`) that will extend the capabilities of the Hedera Smart Contract Service (HSCS) to support frictionless airdrops.
```

Reviewer 4: Spelling mistake

Author: In the case of NFTs we can reject multiple serials for a single token, thus the array argument.

- Expose `cancelAirdropFT`, `cancleAirdropNFT`, `claimAirdropFT`, `claimAirdropNFT`, `rejectTokenFt`, `rejectTokenNFTs`, `setUnlimitedAutomaticAssociations` as new functions in a new proxy redirect token facade contract IHRC904.
- Implement the needed HTS system contract classes to support the new functions.

## Non Goals

- The implementation of the HAPI operation, as it is already an existing feature.

## Architecture

The architecture for the frictionless airdrops follows the existing framework defined for handling all calls to the HederaTokenService system contract in the modularization services and is described in more detail in the Implementation section below.

## Implementation

### New Solidity Structures

We will introduce the following new structures to support the new functionality:

`PendingAirdrop` - A struct that represents a pending airdrop request.

```solidity
struct PendingAirdrop {
    address sender;
    address receiver;

    address token;
    int64 serial;
}
```

`NftId` - A struct that represents the Nft serial to be rejected.

```solidity

> struct NftId {

Author: Updated

|------------------------|----------------------------------------------------------------------------------------------------------------------|------------------|---------------------|----------------|---------------------------------|
| `0xcef5b705`           | `function cancelAirdropFT(address receiverAddress) external returns (uint256 responseCode)`                          | TokenCancel      | HTS                 | `ResponseCode` | The response code from the call |
| `0xad4917cf`           | `function cancelAirdropNFT(address receiverAddress, int64 serialNumber) external returns (uint256 responseCode)`     | TokenCancel      | HTS                 | `ResponseCode` | The response code from the call |
| `0xa83bc5b2`           | `function claimAirdropFT(address senderAddress) external returns (uint256 responseCode)`                             | TokenClaim       | HTS                 | `ResponseCode` | The response code from the call |
| `0x63ada5d7`           | `function claimAirdropNFT(address senderAddress, int64 serialNumber) external returns (uint256 responseCode)`        | TokenClaim       | HTS                 | `ResponseCode` | The response code from the call |

> | `0x76c6b391`           | `function rejectTokenFT() external returns (uint256 responseCode)`                                                   | TokenReject      | HTS                 | `ResponseCode` | The response code from the call |

Author: As this is a token redirect call there is no need for arguments in this case. We use {msg.sender} to get the account address and the redirected token itself to construct the TokenReject transaction body.
